bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java

```diff
@@ -178,16 +130,18 @@ protected void addItemFactory(ItemFactory itemFactory) {
 
         if (failedToCreate.size() > 0) {
             // retry failed creation attempts
-            Iterator<ItemInfo> iterator = failedToCreate.iterator();
+            Iterator<String> iterator = failedToCreate.keySet().iterator();
             while (iterator.hasNext()) {
-                ItemInfo itemInfo = iterator.next();
-                Item item = itemFactory.createItem(itemInfo.itemType, itemInfo.itemName);
+                String itemName = iterator.next();
+                PersistedItem persistedItem = failedToCreate.get(itemName);
+                ActiveItem item = itemFactory.createItem(persistedItem.itemType, itemName);
                 if (item != null) {
-                    iterator.remove();
+                    failedToCreate.remove(itemName);
```

Comment:
You are right: <code>ConcurrentHashMap</code>
Returns a Set view of the keys contained in this map. The set is backed by the map, so changes to the map are reflected in the set, and vice-versa. The set supports element removal, which removes the corresponding mapping from this map, via the Iterator.remove, Set.remove, removeAll, retainAll, and clear operations. It does not support the add or addAll operations.


Shouldn't we implement against the map interface if we declarate the failedToCreate this way <code>Map&lt;String, PersistedItem&gt; failedToCreate</code>?
Sure, ConcurrentHashMap is used for the initialization, but for stability / readability in a few month I would prefer using iterator.remove that is working for the interface and the special implementation.